text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+`click` handler? Which update method should I use - `didRender`, `didUpdate`, `didReceiveAttrs`?
+
+Glimmer components have a constructor, **3** lifecycle hooks, and **1** property
+(`args`). They don't extend from any classes, themselves being a simple native
+class. They don't have any element/DOM based properties, hooks, event handler
+functions, whose responsibilities have been passed on to element modifiers. This
+_dramatically_ simplifies what users need to learn in order to start using the
+bread-and-butter class of Ember, and once again encourages _thin_, template
+oriented components.
+
+### Glimmer.js Compatibility
+
+One of the goals for future versions of Ember, post Ember Octane, will be to
+enable lighter-weight applications to be built using the framework. Breaking
+Ember apart into smaller, fully independent and optional pieces is the core idea
+behind the "install your way to Ember" goal, which will enable Ember to be used
+in more constrained environments that smaller frameworks such as React, Preact,
+Vue, and more excel in. It will also allow users who are size-conscious to adopt
+Ember incrementally, adding functionality when it is _needed_ rather than having
+all-or-nothing.
+
+This will take time though. Much progress has been made, but parts of Ember are
+still very monolithic. And while it isn't Ember, Glimmer.js is a _very_
+lightweight wrapper of the Glimmer VM that enables users to drop that weight and
+begin writing much more minimal apps today.
+
+Glimmer components aren't just based on the Glimmer.js component API - they are
+one and the same. They will be a shared package, which will be importable and
+usable by the users of both frameworks. Not only will users be able to write
+better components in Ember, those components will also be cross-compatible with
+Glimmer apps (assuming they don't use Ember specific functionality).
+
+It is important to note that while Glimmer components will be versioned
+independently from Glimmer and Ember ***they will abide by the Ember RFC process
+for any and all changes to user APIs***. The implementations for their component
+managers in Glimmer and Ember may change to keep them compatible, but they will
+not make major changes without first getting community input, and will be
+considered part of the public API of Ember.
+
+## Prior Art
+
+* [Glimmer.js](https://glimmerjs.com/) is the framework that this component API
+  is based on, and will be cross-compatible with.
+* [sparkles-component](https://github.com/rwjblue/sparkles-component) is an
+  an implementation of the Glimmer.js component API using Ember's component
+  managers. It is usable in Ember today.
+
+## Detailed design
+
+Glimmer components have the following API:
+
+```ts
+class GlimmerComponent<Args> {
+  constructor(injections: object, public args: Args);
+
+  didCreate(): void;
+  didUpdate(): void;
+
+  destroy(): void;
+}
+```
+
+In addition, two new element modifiers will fill in the gaps left by the
```

Comment:
I'm assuming that element modifiers can be applied to other glimmer components, ala:

```
<parent-component>
  <child-component {{on-insert this.childDidUpdate}} />
<parent-component>
```

Reply:
I don't believe so, based on the current [modifiers RFC](https://github.com/emberjs/rfcs/pull/353/files#diff-8b3eda7c0c2bc69ef34db57bd02fd749R248), it looks like they can only be invoked on actual elements and will result in a compiler error if used on components.